### PR TITLE
[feg] Add GxGy disable and Charging for Omnipresent rules

### PR DESCRIPTION
**Pull request by uri200**
Signed-off-by: Oriol Batalla <obatalla@fb.com>

## Summary

This PR includes:
- Omnipresent rules with charging keys: the way we achieve that is injecting omni rules and omni base rule names to the CCA-I received by the PCRF. Note OCS needs tho have provisioned those charging keys in order to record quota.

- Disable Gx/Gy through configuration: added a new configuration parameter to indicate if Gx or Gy is disabled. In case it is disabled we do the following
	- Set disabled on config mconfig parameter 
	- Do not create the Gx or Gy client during execution of main in case of any of them disabled
	- Do not send createRequest for Gx or Gy in case of disabled
	- In case of Gx disabled we generate an empty CCA-I response so we can keep injecting omnirules even if Gx is 		disabled

Note the important changes can be seen at
`session_controller.go`

Then the changes on config for the addition of this parameters can be seen at
`main.go`
`gx_config.go`
`gy_config.go`
`mconfig.go`

## Test Plan

orcr8r 
	`./build -g `
	`./build -m make precommit`
feg 
	`make precommit`
cwag_integ_test

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: cwf/gateway/integ_tests/gy_enforcement_test.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func ocsCreditExhaustionTestSetup(t *testing.T) (*TestRunner, *RuleManager, *cwfprotos.UEConfig) {
+func ocsAndPcrfCreditExhaustionTestSetup(t *testing.T) (*TestRunner, *RuleManager, *cwfprotos.UEConfig) {
```
Comment (uri200) [inline cwf/gateway/integ_tests/gy_enforcement_test.go:38]: I changed the name because it was hard for me to find where the PCRF was actually configured

File: feg/cloud/go/plugin/mconfig.go [hunk 2]
```diff
@@ -238,3 +240,7 @@ func getGyInitMethod(initMethod *uint32) mconfig.GyInitMethod {
 	}
 	return mconfig.GyInitMethod(*initMethod)
 }
+
+func getBooleanFromPtr(b *bool) bool {
+	return *b
```
Comment (themarwhal) [inline feg/cloud/go/plugin/mconfig.go]: We use `swag.BoolValue` in orc8r for these kinds of things

File: feg/cloud/go/plugin/mconfig_test.go [hunk 3]
```diff
@@ -389,3 +393,7 @@ var defaultConfig = &models.NetworkFederationConfigs{
 func uint32Ptr(i uint32) *uint32 {
 	return &i
 }
+
+func boolPtr(b bool) *bool {
+	return &b
```
Comment (themarwhal) [inline feg/cloud/go/plugin/mconfig_test.go]: `swag.Bool` might be useful here too

File: feg/gateway/services/session_proxy/servicers/session_controller.go [hunk 4]
```diff
@@ -172,15 +171,15 @@ func (srv *CentralSessionController) handleUseGyForAuthOnly(
 	}, nil
 }
 
-func (srv *CentralSessionController) getChargingKeysFromRuleInstalls(
+func getChargingKeysFromRuleInstalls(
```
Comment (themarwhal) [inline feg/gateway/services/session_proxy/servicers/session_controller.go]: Why this change?

File: feg/gateway/services/session_proxy/servicers/session_controller_test.go [hunk 5]
```diff
@@ -370,7 +370,9 @@ func standardUsageTest(
 			policydb.ChargingKey{RatingGroup: 21}}, nil).Once()
 	// no omnipresent rules
 	mockPolicyDb.On("GetOmnipresentRules").Return([]string{}, []string{}).Once()
-	mockPolicyDb.On("GetRuleIDsForBaseNames", mock.Anything).Return([]string{}).Once()
+
+	//DELME
```
Comment (themarwhal) [inline feg/gateway/services/session_proxy/servicers/session_controller_test.go]: DELME!

File: feg/gateway/services/session_proxy/servicers/session_controller_test.go [hunk 6]
```diff
@@ -1295,6 +1355,7 @@ func initMconfig() error {
                }
             },
             "gy": {
+			   "disableGy": false,
```
Comment (themarwhal) [inline feg/gateway/services/session_proxy/servicers/session_controller_test.go:1356]: fmt?

File: cwf/gateway/integ_tests/omni_rules_test.go [hunk 7]
```diff
@@ -52,37 +55,62 @@ func TestOmnipresentRules(t *testing.T) {
...
+	gyInitAnswer:= protos.NewGyCCAnswer(diam.Success).SetQuotaGrant(quotaGrant)
+	gyInitExpectation := protos.NewGyCreditControlExpectation().Expect(gyInitRequest).Return(gyInitAnswer)
+	gyExpectations := []*protos.GyCreditControlExpectation{gyInitExpectation}
+	assert.NoError(t, setOCSExpectations(gyExpectations, nil))
 
-	tr.AuthenticateAndAssertSuccessWithRetries(imsi, 5)
+	tr.AuthenticateAndAssertSuccess(imsi)
 
-	req := &cwfprotos.GenTrafficRequest{Imsi: imsi, Volume: &wrappers.StringValue{Value: *swag.String("200k")}}
+	req := &cwfprotos.GenTrafficRequest{Imsi: imsi, Volume: &wrappers.StringValue{Value: *swag.String("200K")}}
```
Comment (themarwhal) [inline cwf/gateway/integ_tests/omni_rules_test.go]: Not your change, but *swag.String("200K") is equivalent to "200K", I think

**Review by themarwhal — APPROVED** [review]

File: feg/gateway/services/session_proxy/servicers/session_controller.go (hunk 4 above)
Comment (uri200) [reply feg/gateway/services/session_proxy/servicers/session_controller.go]: Oh, i forgot that one. 

I started to remove some functions that were attached to the session controller type. Then I didn't like it after reading about functions vs methods on go and I put them back. I forgot that one. 

I will put it back again as it was

File: cwf/gateway/integ_tests/omni_rules_test.go (hunk 7 above)
Comment (uri200) [reply cwf/gateway/integ_tests/omni_rules_test.go]: oh that swag.String thing was there before. I didnt even questioned it wasn't needed. I was just playing with the values. I will change that

File: feg/gateway/services/session_proxy/servicers/session_controller_test.go (hunk 5 above)
Comment (uri200) [reply feg/gateway/services/session_proxy/servicers/session_controller_test.go]: 🤦‍♂️

File: feg/gateway/services/session_proxy/servicers/session_controller_test.go (hunk 6 above)
Comment (uri200) [reply feg/gateway/services/session_proxy/servicers/session_controller_test.go:1356]: Do you mean format? I don't see any issue there.

image

Comment (themarwhal) [reply feg/gateway/services/session_proxy/servicers/session_controller_test.go:1356]: Maybe just the github view ... 🤦‍♂️
